storage/src/vespa/storage/distributor/pendingmessagetracker.h

```diff
@@ -147,48 +160,167 @@ class PendingMessageTracker : public framework::HtmlStatusReporter {
-        >
-    >;
-
-    // Must match Messages::nth_index<N>
-    static constexpr uint32_t IndexByMessageId     = 0;
-    static constexpr uint32_t IndexByNodeAndBucket = 1;
-    static constexpr uint32_t IndexByBucketAndType = 2;
+    // wraps an iterator for std::set<MessageEntry *>, hiding the extra indirection
+    template<typename I> struct wrap_set_iterator : public I {
+        auto * operator-> () const { return I::operator*(); }
+        auto & operator* () const { return *I::operator*(); }
+    };
+
+    // multi-index container:
+    struct Messages {
+        using MainSet = std::set<MessageEntry, MessageIdKey>;
+        using ByNodeAndBucketSet = std::set<const MessageEntry *, NodeBucketTypeIdComparator>;
+        using ByBucketAndTypeSet = std::set<const MessageEntry *, BucketTypeNodeIdComparator>;
+
+        // generic emplace, keeping indexes in sync
+        template<typename ...Args>
+        void emplace(Args&&... args)
+        {
+            auto [iter, added] = byMessageIdSet.emplace(std::forward<Args>(args)...);
+            assert(added);
+            const MessageEntry &entry = *iter;
+            byNodeAndBucketSet.insert(&entry);
+            byBucketAndTypeSet.insert(&entry);
+        }
+
+        // remove by key, keeping indexes in sync
+        void remove(const MessageEntry &key) {
+            auto iter = byMessageIdSet.find(key);
+            assert (iter != byMessageIdSet.end());
+            const MessageEntry &entry = *iter;
+            byNodeAndBucketSet.erase(&entry);
+            byBucketAndTypeSet.erase(&entry);
+            byMessageIdSet.erase(iter);
+        }
+
+        // Index wrapping the set using NodeBucketTypeIdComparator
+        struct IndexByNodeAndBucket {
+            using iterator = wrap_set_iterator<ByNodeAndBucketSet::const_iterator>;
+            Messages& _m;
+            auto begin() const { return iterator(_m.byNodeAndBucketSet.begin()); }
+            auto end()   const { return iterator(_m.byNodeAndBucketSet.end()); }
+
+            IndexByNodeAndBucket(Messages& m) : _m(m) {}
+            template<typename Key>
+            std::pair<iterator, iterator> equal_range(Key key) const {
+                auto inner_range = _m.byNodeAndBucketSet.equal_range(key);
+                return std::make_pair(iterator(inner_range.first),
+                                      iterator(inner_range.second));
+            }
+            iterator erase(iterator it) {
+                const MessageEntry& entry = *it;
+                ++it;
+                _m.remove(entry);
+                return it;
+            }
+            void erase(iterator it, iterator to) {
+                while (it != to) {
+                    it = erase(it);
+                }
+            }
+        };
+
+        // Index wrapping the set using BucketTypeNodeIdComparator
+        struct IndexByBucketAndType {
+            using iterator = wrap_set_iterator<ByBucketAndTypeSet::const_iterator>;
```

Comment:
Should `iterator` in `IndexByBucketAndType` and `IndexByNodeAndBucket` technically be called `const_iterator` instead since they're returned from `const` overloads?

Reply:
agreed, renamed these to `const_iterator`